src/main/java/org/broadinstitute/hellbender/utils/bwa/BwaMemIndex.java

```diff
@@ -0,0 +1,182 @@
+        }
+        ByteBuffer refContigNamesBuf = getRefContigNames(indexAddress);
+        if ( refContigNamesBuf == null ) {
+            throw new IllegalStateException("Unable to retrieve reference contig names from bwa-mem index "+indexImageFile);
+        }
+        refContigNamesBuf.order(ByteOrder.nativeOrder()).position(0).limit(refContigNamesBuf.capacity());
+        int nRefContigNames = refContigNamesBuf.getInt();
+        refContigNames = new ArrayList<>(nRefContigNames);
+        for ( int idx = 0; idx < nRefContigNames; ++idx ) {
+            int nameLen = refContigNamesBuf.getInt();
+            byte[] nameBytes = new byte[nameLen];
+            refContigNamesBuf.get(nameBytes);
+            refContigNames.add(new String(nameBytes));
+        }
+        destroyByteBuffer(refContigNamesBuf);
+    }
+
+    public boolean isOpen() { return indexAddress != 0L; }
+
+    public long refIndex() {
+        refCount.incrementAndGet();
+        if ( indexAddress == 0L ) {
+            throw new IllegalStateException("Index image " + indexImageFile + " has been closed");
+        }
+        return indexAddress;
+    }
+
+    public void deRefIndex() { refCount.decrementAndGet(); }
+
+    @Override
+    public void close() {
+        long addr = indexAddress;
+        if ( addr != 0L ) {
+            synchronized (BwaMemIndex.class) {
+                addr = indexAddress;
+                if ( addr != 0L ) {
+                    if ( refCount.intValue() != 0 ) {
+                        throw new IllegalStateException("Index image "+indexImageFile+" can't be closed:  it's in use.");
+                    }
+                    indexAddress = 0L;
+                    destroyIndex(indexAddress);
+                }
+            }
+        }
+    }
+
+    public List<String> getReferenceContigNames() {
+        return refContigNames;
+    }
+
+    public static String getBWAVersion() {
+        loadNativeLibrary();
+        return getVersion();
+    }
+
+    ByteBuffer doAlignment( final ByteBuffer seqs, final ByteBuffer opts ) {
+        final ByteBuffer alignments = createAlignments(seqs, indexAddress, opts);
+        if ( alignments == null ) {
+            throw new IllegalStateException("Unable to get alignments from bwa-mem index "+indexImageFile+": We don't know why.");
+        }
+        return alignments;
+    }
+
+    private static void assertNonEmptyReadable( final String fileName ) {
+        if ( !nonEmptyReadableFile(fileName) )
+            throw new IllegalArgumentException("Missing bwa index file: "+fileName);
+    }
+
+    private static boolean nonEmptyReadableFile( final String fileName ) {
+        try ( final FileInputStream is = new FileInputStream(fileName) ) {
+            return is.read() != -1;
+        } catch ( final IOException ioe ) {
+            return false;
+        }
+    }
+
+    private static void loadNativeLibrary() {
+        if ( !nativeLibLoaded ) {
+            synchronized(BwaMemIndex.class) {
```

Comment:
I'd like to move this logic to some tiny library somewhere so people don't have to keep implementing it....